src/v/cloud_topics/objects/level_one_object.h

```diff
@@ -0,0 +1,392 @@
+
+    object_index copy() const;
+
+    auto serde_fields() { return std::tie(partitions); }
+
+    bool operator==(const object_index&) const = default;
+
+    // The value returned when an index search doesn't have contain matching
+    // data.
+    constexpr static size_t npos = std::numeric_limits<size_t>::max();
+
+    // Return the file position of the latest record batch that has the offset
+    // at or before the given offset. If the offset is not in this file then
+    // `npos` is returned.
+    //
+    // Example:
+    //
+    // If the footer has the following offset ranges indexed for the given
+    // ntp:
+    //
+    // [[1, 10], [11, 20], [30, 40]]
+    //
+    // Searching for offset 5 would yield the position of the batch[0],
+    // while a search for offsets 25 or 40 would yield batch[2]. Searching for
+    // offset 50 would yield `npos`.
+    size_t file_position_before_kafka_offset(const model::ntp&, kafka::offset);
+
+    // Return the file position of the latest record batch that has a
+    // max_timestamp at or before the given timestamp. If the timestamp is
+    // greater than all timestamps in this file, then `npos` is returned.
+    //
+    // Example:
+    //
+    // If the footer has the following max timestamps indexed for the given
+    // ntp:
+    //
+    // 3, 10, 10, 10, 40
+    //
+    // Searching for timestamp 5 or 10 would yield the position of the batch[1],
+    // the timestamp 1 would yield the position of batch[0].
+    // While a search for timestamp 25 or 40 would yield batch[4] and the
+    // timestamp 50 would yield `npos`.
+    size_t
+    file_position_before_max_timestamp(const model::ntp&, model::timestamp);
+
+    // Read the object_index using the suffix of an L1 object.
+    //
+    // Returns either the object_index, or the *additional* bytes needed to be
+    // prepended to the iobuf in order to complete reading the footer.
+    //
+    // REQUIRES: that the iobuf is at least the last 4 bytes of the file, but
+    // likely you want to optimistically read more data (say 512KiB) if you
+    // don't know the the exact footer location.
+    //
+    // Here's an example of how to use this function:
+    //
+    // ```c++
+    // size_t object_size = ...;
+    // auto iobuf = read_object(
+    //   handle,
+    //   {.offset = object_size - 1_KiB, .size = 1_KiB},
+    // );
+    // auto result = l1::object_index::read_footer(iobuf.share());
+    // if (std::holds_alternative<l1::object_index>(result)) {
+    //   return std::get<l1::object_index>(result);
+    // }
+    // size_t extra = std::get<size_t>(result);
+    // auto missing = read_object(
+    //   handle,
+    //   {.offset = object_size - 1_KiB - extra, .size = extra},
+    // );
+    // missing.append(std::move(iobuf));
+    // result = l1::object_index::read_footer(std::move(missing));
+    // return std::get<l1::object_index>(result);
+    // ```
+    static std::variant<object_index, size_t> read_footer(iobuf);
```

Comment:
I'm wondering if footer should have its own serde type (to make it easy to extend in the future). Currently footer == object_index

Reply:
I just renamed this to footer, I don't see it being worthwhile to add another layer of struct.

Done in https://github.com/redpanda-data/redpanda/pull/26563/commits/7962c0738615826bf1cea013915362551bdd38a5